core/services/llo/telemetry.go

```diff
@@ -81,28 +86,116 @@ func (t *telemeter) EnqueueV3PremiumLegacy(run *pipeline.Run, trrs pipeline.Task
 		// ignore errors
 		return
 	}
-	tObs := TelemetryObservation{run, trrs, streamID, opts, val, dpInvariantViolationDetected}
+	tp := &TelemetryPipeline{run, trrs, streamID, opts, val, dpInvariantViolationDetected}
 	select {
-	case t.chTelemetryObservation <- tObs:
+	case t.chTelemetryPipeline <- tp:
 	default:
 	}
 }
 
+type telemetryCollectionContext struct {
+	in   <-chan interface{}
+	opts llo.DSOpts
+}
+
+// CollectTelemetryObserve reads telem packets from the returned channel and
+// sends them to the monitoring endpoint. Stops reading when channel closed or
+// when telemeter is stopped
+//
+// CALLER IS RESPONSIBLE FOR CLOSING THE RETURNED CHANNEL TO AVOID MEMORY
+// LEAKS.
+func (t *telemeter) MakeTelemChannel(opts llo.DSOpts, size int) chan<- interface{} {
+	ch := make(chan interface{}, size)
+	tcc := telemetryCollectionContext{
+		in:   ch,
+		opts: opts,
+	}
+
+	select {
+	case t.chch <- tcc:
+	default:
+		// This should be performant enough with buffer of t.chch large enough
+		// that we never hit this case, however, we should NEVER block
+		// observations on telemetry even if something pathological happens.
+		t.eng.SugaredLogger.Errorw("Telemeter chch full, will not record telemetry", "seqNr", opts.SeqNr())
```

Comment:
The logger is embedded:
```suggestion
		t.eng.Errorw("Telemeter chch full, will not record telemetry", "seqNr", opts.SeqNr())
```

Reply:
Cool, didn't know this!